crates/oauth2-types/src/oidc.rs

```diff
@@ -233,3 +250,541 @@ pub struct Metadata {
+            Some(u) => u,
+            None => unreachable!(),
+        }
+    }
+
+    /// URL of the authorization server's token endpoint.
+    #[must_use]
+    pub fn token_endpoint(&self) -> &Url {
+        match &self.token_endpoint {
+            Some(u) => u,
+            None => unreachable!(),
+        }
+    }
+
+    /// URL of the authorization server's JWK Set document.
+    #[must_use]
+    pub fn jwks_uri(&self) -> &Url {
+        match &self.jwks_uri {
+            Some(u) => u,
+            None => unreachable!(),
+        }
+    }
+
+    /// JSON array containing a list of the OAuth 2.0 `response_type` values
+    /// that this authorization server supports.
+    #[must_use]
+    pub fn response_types_supported(&self) -> &[OAuthAuthorizationEndpointResponseType] {
+        match &self.response_types_supported {
+            Some(u) => u,
+            None => unreachable!(),
+        }
+    }
+
+    /// JSON array containing a list of the Subject Identifier types that this
+    /// OP supports.
+    #[must_use]
+    pub fn subject_types_supported(&self) -> &[SubjectType] {
+        match &self.subject_types_supported {
+            Some(u) => u,
+            None => unreachable!(),
+        }
+    }
+
+    /// JSON array containing a list of the JWS `alg` values supported by the OP
+    /// for the ID Token.
+    #[must_use]
+    pub fn id_token_signing_alg_values_supported(&self) -> &[JsonWebSignatureAlg] {
+        match &self.id_token_signing_alg_values_supported {
+            Some(u) => u,
+            None => unreachable!(),
+        }
+    }
+}
+
+impl Deref for VerifiedProviderMetadata {
+    type Target = ProviderMetadata;
+
+    fn deref(&self) -> &Self::Target {
+        &self.inner
+    }
+}
+
+/// All errors that can happen when verifying [`ProviderMetadata`]
+#[derive(Debug, Error)]
+pub enum ProviderMetadataVerificationError {
+    /// The issuer is missing.
+    #[error("issuer is missing")]
+    MissingIssuer,
+
+    /// The authorization endpoint is missing.
+    #[error("authorization endpoint is missing")]
+    MissingAuthorizationEndpoint,
+
+    /// The token endpoint is missing.
+    #[error("token endpoint is missing")]
+    MissingTokenEndpoint,
+
+    /// The JWK Set URI is missing.
+    #[error("JWK Set URI is missing")]
+    MissingJwksUri,
+
+    /// The supported response types are missing.
+    #[error("supported response types are missing")]
+    MissingResponseTypesSupported,
+
+    /// The supported subject types are missing.
+    #[error("supported subject types are missing")]
+    MissingSubjectTypesSupported,
+
+    /// The supported ID token signing algorithm values are missing.
+    #[error("supported ID token signing algorithm values are missing")]
+    MissingIdTokenSigningAlgValuesSupported,
+
+    /// The URL of the given field doesn't use a `https` scheme.
+    #[error("{0}'s URL doesn't use a https scheme: {1}")]
+    UrlNonHttpsScheme(String, Url),
```

Comment:
Those strings are always known at compile time, right? They could be `&'static str` then, avoiding some `to_owned` calls here and there